crates/sparrow-compiler/src/ast_to_dfg.rs

```diff
@@ -527,31 +527,28 @@ pub(super) fn add_to_dfg(
                 dfg.bind("$condition_input", args[0].inner().clone());
 
                 let window = &expr.unwrap().args()[1];
-                let (condition, duration) = match window.op() {
-                    ExprOp::Call(window_name) => {
-                        flatten_window_args(window_name, window, dfg, data_context, diagnostics)?
-                    }
-                    ExprOp::Literal(v) if v.inner() == &LiteralValue::Null => {
-                        // Unwindowed aggregations just use nulls
-                        let null_arg = dfg.add_literal(LiteralValue::Null.to_scalar()?)?;
-                        let null_arg = Located::new(
-                            add_literal(
-                                dfg,
-                                null_arg,
-                                FenlType::Concrete(DataType::Null),
-                                window.location().clone(),
-                            )?,
-                            window.location().clone(),
-                        );
-
-                        (null_arg.clone(), null_arg)
-                    }
-                    unexpected => anyhow::bail!("expected window, found {:?}", unexpected),
-                };
+                let (condition, duration) =
+                    flatten_window_args_if_needed(window, dfg, data_context, diagnostics)?;
```

Comment:
Thought for the future: One nice thing about the Python route -- I think we can make this a *lot* cleaner. What I'm imagining is that Sparrow exposes more specific functions for kinds of windowed aggregation, and then this flattening / expansion can happen within Python. But -- we can use *Python types* to manage windows and such.

Consider something like the following. This is nice because we don't need to encode the "special" (builder-only) types within Sparrow. 

```
def sum(input: Expr, window: Window = None) -> Expr:
  if window is None:
    return Expr.call("sum", input)
  elif isinstance(window, SinceWindow):
    return Expr.call("sum_windowed", input, window.predicate)
  else:
    pass
```

Reply:
Yeah, that looks like it helps with managing window type sin the python side. The other side of things is how the window functions like `daily()` need to know the entity domain from the `input`. If we make it such that functions have windows built-in such that they're inherently part of the same domain then that would get rid of the `env` binding we have to do in `ast_to_dfg`.